### ISSUE TITLE
Image.toByteData() results in incorrect data

### ISSUE THREAD

**Issue by Jaween** [flutter/flutter#26515]
Calling `dart:ui::Image.toByteData()` results in a buffer which doesn't always match the image exactly.

#### Minimal example
```dart
import 'dart:ui' as Ui;
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final image = _drawImage();
    _printImageBuffer(image);
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: CustomPaint(
            size: Size(200, 200),
            painter: MyCustomPainter(image),
          ),
        ),
      ),
    );
  }

  Ui.Image _drawImage() {
    final path = Path();
    path.moveTo(0.0, 0.0);
    path.lineTo(3.5, 3.5);

    final paint = Paint();
    paint.style = PaintingStyle.stroke;
    paint.strokeWidth = 0;
    paint.color = Color.fromARGB(0xFF, 0x11, 0x11, 0x11);
    paint.isAntiAlias = false;

    final pictureRecorder = Ui.PictureRecorder();
    final canvas = Canvas(pictureRecorder);
    canvas.drawPath(path, paint);
    final picture = pictureRecorder.endRecording();
    return picture.toImage(4, 4);
  }

  void _printImageBuffer(Ui.Image image) async {
    final byteData = await image.toByteData();

    final buffer = StringBuffer();
    for (int i = 0; i < byteData.lengthInBytes; i += 4) {
      buffer.write((i / 4) % image.width == 0 ? "\n" : "");
      buffer.write(byteData.getUint8(i) != 0 ? "#" : ".");
    }
    print(buffer.toString());
  }
}

class MyCustomPainter extends CustomPainter {
  final Ui.Image image;

  MyCustomPainter(this.image);

  @override
  void paint(Canvas canvas, Size size) {
    final srcRect = Rect.fromLTWH(0, 0, 4, 4);
    final destRect = Offset.zero & size;
    canvas.drawImageRect(image, srcRect, destRect, Paint());
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}
```

The `_printImageBuffer()` function visualises the buffer as ASCII art simply so that we can observe the discrepancies.

#### Case 1
A path is drawn from `(0.0, 0.0)` to `(3.5. 3.5)`.

Visual output:
<img src="https://user-images.githubusercontent.com/5057145/51099515-22e60c80-1821-11e9-9bde-f738610f47e6.jpg" width="400"/>
Console output:
```
I/flutter (17886): 
I/flutter (17886): #...
I/flutter (17886): .#..
I/flutter (17886): ..#.
I/flutter (17886): ...#
```

The image does not match the ASCII visualisation (an extra `#` in the lower right).

#### Case 2
A path is drawn from `(0.0, 0.0)` to `(4.0. 4.0)`.

Visual output:
<img src="https://user-images.githubusercontent.com/5057145/51099521-2d080b00-1821-11e9-8996-b785103892f8.jpg" width="400"/>
Console output:
```
I/flutter (17886): 
I/flutter (17886): #...
I/flutter (17886): .#..
I/flutter (17886): ..#.
I/flutter (17886): ...#
```

The image matches the ASCII visualisation.

#### Flutter Doctor output
```
[✓] Flutter (Channel master, v1.1.10-pre.44, on Linux, locale en_AU.UTF-8)
    • Flutter version 1.1.10-pre.44 at /opt/flutter
    • Framework revision 6c7fb88059 (2 days ago), 2019-01-11 23:18:48 -0500
    • Engine revision 5722a9685e
    • Dart version 2.1.1 (build 2.1.1-dev.1.0 f0c7d971c4)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /home/jaween/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)
    • All Android licenses accepted.

[✓] Android Studio (version 3.2)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 31.3.1
    • Dart plugin version 181.5656
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)

[✓] VS Code (version 1.29.1)
    • VS Code at /usr/share/code
    • Flutter extension version 2.21.1

[✓] Connected device (1 available)
    • SM A520F • 5200676a4a815579 • android-arm64 • Android 8.0.0 (API 26) (emulator)

• No issues found!
```

**Comment by Jaween** [comment]
Running the app using `flutter run --enable-software-rendering` appears to fix this issue in my test. It may be related to #23621.

**Comment by Jaween** [comment]
The fix for #23621 has indeed seemed to have fixed this issue :)

Since `Picture.toImage()` is now returns a `Future`, I've had to updated the example from the original post. Here it is for reference:

```dart
import 'dart:ui' as Ui;
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  MyAppState createState() {
    return new MyAppState();
  }
}

class MyAppState extends State<MyApp> {

  Ui.Image image;

  @override
  void initState() {
    super.initState();
    setup();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: CustomPaint(
            size: Size(200, 200),
            painter: MyCustomPainter(image),
          ),
        ),
      ),
    );
  }

  void setup() async {
    final tempImage = await _drawImage();
    setState(() => image = tempImage);
    _printImageBuffer(tempImage);
  }

  Future<Ui.Image> _drawImage() {
    final path = Path();
    path.moveTo(0.0, 0.0);
    path.lineTo(3.5, 3.5);

    final paint = Paint();
    paint.style = PaintingStyle.stroke;
    paint.strokeWidth = 0;
    paint.color = Color.fromARGB(0xFF, 0x11, 0x11, 0x11);
    paint.isAntiAlias = false;

    final pictureRecorder = Ui.PictureRecorder();
    final canvas = Canvas(pictureRecorder);
    canvas.drawPath(path, paint);
    final picture = pictureRecorder.endRecording();
    return picture.toImage(4, 4);
  }

  void _printImageBuffer(Ui.Image image) async {
    final byteData = await image.toByteData();

    final buffer = StringBuffer();
    for (int i = 0; i < byteData.lengthInBytes; i += 4) {
      buffer.write((i / 4) % image.width == 0 ? "\n" : "");
      buffer.write(byteData.getUint8(i) != 0 ? "#" : ".");
    }
    print(buffer.toString());
  }
}

class MyCustomPainter extends CustomPainter {
  final Ui.Image image;

  MyCustomPainter(this.image);

  @override
  void paint(Canvas canvas, Size size) {
    final srcRect = Rect.fromLTWH(0, 0, 4, 4);
    final destRect = Offset.zero & size;
    if (image != null) {
      canvas.drawImageRect(image, srcRect, destRect, Paint());
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}
```

**Comment by dkashkin** [comment]
Thank you @Jaween! You were right - I see this in Master. Appreciate your help!
